### PR TITLE
Customdocker

### PR DESCRIPTION
**Pull request by mcvidanagama**

### CODE REVIEW

File: modules/jaggeryapps/appmgt/src/site/themes/default/templates/home/js/home.js [hunk 1]
```diff
@@ -76,7 +76,7 @@ function generateDefaultLaunchUrl() {
 
 function loadEndpointView() {
     clearInterval(timerId);
-    if (selectedApplicationRevision.status == APPLICATION_RUNNING) {
+    if (selectedApplicationRevision.status == APPLICATION_RUNNING && application.applicationType != custom) { // endpoints will not be loaded and displayed for custom docker image applications
```
Comment (amalkasubasinghe) [inline modules/jaggeryapps/appmgt/src/site/themes/default/templates/home/js/home.js]: don't we require adding quotes around custom

File: modules/jaggeryapps/appmgt/src/site/themes/default/templates/application/template.jag [hunk 2]
```diff
@@ -1258,8 +1285,26 @@ jagg.template("application", function (inputs, outputs, jagg) {
...
+            <%} else {%>
+             <div class="form-group" id="custom-docker-image-block">
+                <div class="input-group">
+                    <span class="input-group-addon">Image ID</span>
+                    <input id="remoteImageUrl" name="remoteImageUrl" type="text" class="form-control" placeholder="docker.io/testuser/testimage:testtag" />
+                </div>
+                </br>
+               <!-- <div class="input-group">
+                    <span class="input-group-addon">User name</span>
+                    <input id="userName" name="userName" type="text" class="form-control" placeholder="required only if repo is private" />
```
Comment (amalkasubasinghe) [inline modules/jaggeryapps/appmgt/src/site/themes/default/templates/application/template.jag]: required 'R' should be capital

File: modules/jaggeryapps/appmgt/src/site/themes/default/templates/application/template.jag [hunk 3]
```diff
@@ -775,28 +794,31 @@ jagg.template("application", function (inputs, outputs, jagg) {
...
-                        && $("#gitRepoBranch").val()))) {// checks form for validity
 
-            if($("#uploadedFileName").val() && progress == 100){
-                if(validateArtifactType($("#uploadedFileName").val())){
+        //first disable the button and then enable it only if condions are satisfied
+        $("#add-application").prop("disabled", true);
+
+        if ($('#applicationName').val() && $('#applicationRevision').val() && $('#container option:selected').val() &&
+            $('#runtime').val() > 0 && checkApplicationVersionExistence($('#applicationRevision').val())) {// mandatory validations for all app types
+            if (appTypeName == custom) { // validations for custom app type
```
Comment (amalkasubasinghe) [inline modules/jaggeryapps/appmgt/src/site/themes/default/templates/application/template.jag]: don't we require quotes around custom

File: modules/jaggeryapps/appmgt/src/modules/application/application.jag [hunk 4]
```diff
@@ -92,17 +93,38 @@ var addApplication = function (applicationName, versionName, applicationDescript
...
-        if (pushDockerImageStatus) {
-            triggerKubernetesDeployment(runtimeProperties, applicationRuntime, tagName, applicationName, versionName, appTypeName, runtimeId, replicas, versionHashId, conSpecMemory, conSpecCpu);
-            putSelectedRevisionToSession(applicationHashId, versionName);
+    var imageName;
+    if (appTypeName != CUSTOM) { //for custom images uploadedFIleName=null
+        var absFileUploadedPath = createDockerFile(applicationName, versionName, appTypeName, runtimeId, uploadedFileName,
+                                                   versionHashId, dockerClient, artifactUrl, hostName, appCreationMethod,
+                                                   runtimeProperties, conSpecMemory);
+        buildDockerImage(applicationName, versionName, runtimeId, absFileUploadedPath, applicationRuntime, tagName,
+                         versionHashId, dockerClient);
```
Comment (amalkasubasinghe) [inline modules/jaggeryapps/appmgt/src/modules/application/application.jag]: Here you have misssed some logics when merging
if build docker images success -> then we do push docker image
if push docker image success -> then we do trigger k8s deployment, etc...

File: modules/dbscripts/appcloud.sql [hunk 5]
```diff
@@ -85,7 +86,8 @@ INSERT INTO `AC_RUNTIME` (`id`, `name`, `image_name`, `tag`, `description`) VALU
 (7, 'WSO2 Data Services Server - 3.5.0', 'wso2dataservice', '3.5.0', 'OS:alpine-java, Oracle JDK:8u102'),
 (8, 'OpenJDK 8 + WSO2 MSF4J 2.0.0', 'msf4j', '2.0.0', 'OS:alpine-java, Oracle JDK:8u102'),
 (9, 'WSO2 Enterprise Service Bus - 5.0.0', 'wso2esb', '5.0.0', 'OS:Debian, Oracle JDK:8u102'),
-(10, 'Apache Tomcat 8.0.28 / WSO2 Application Server 6.0.0-M3', 'wso2as', '6.0.0-m3', 'OS:alpine-java, Oracle JDK:8u102');
+(10, 'Apache Tomcat 8.0.28 / WSO2 Application Server 6.0.0-M3', 'wso2as', '6.0.0-m3', 'OS:alpine-java, Oracle JDK:8u102'),
+(11, 'Custom Docker Image runtime', 'custom', 'customtag', 'OS:Custom, JAVA Version:custom');
```
Comment (amalkasubasinghe) [inline modules/dbscripts/appcloud.sql]: 'OS:Custom, JAVA Version:custom' -> Java Version not applicable here. Can we say something else?

File: modules/components/org.wso2.appcloud.core/src/main/java/org/wso2/appcloud/core/docker/DockerClient.java [hunk 6]
```diff
@@ -232,24 +235,92 @@ public void onEvent(String event) {
...
+            log.error("Docker image pull failed : " + imageRepoUrl + " with tag : " + imageTag);
+            throw new AppCloudException("Docker image pull failed: " + imageRepoUrl + " with tag : " + imageTag);
         }
     }
 
+    public void tagDockerImage(String oldImage, String oldTag, String newImageName, String newTag)
+            throws AppCloudException {
+
+        if (log.isDebugEnabled()) {
+            log.debug("Docker image tag triggered for image : " + oldImage + " with old tag : " + oldTag + " to new " +
```
Comment (amalkasubasinghe) [inline modules/components/org.wso2.appcloud.core/src/main/java/org/wso2/appcloud/core/docker/DockerClient.java]: Docker image tagging triggered for image

File: modules/components/org.wso2.appcloud.core/src/main/java/org/wso2/appcloud/core/docker/DockerClient.java [hunk 7]
```diff
@@ -232,24 +235,92 @@ public void onEvent(String event) {
...
+    }
+
+    public void pullDockerImage(String imageRepoUrl, String imageTag) throws AppCloudException {
+
+        if (log.isDebugEnabled()) {
+            log.debug("Docker image pull triggered for repo : " + imageRepoUrl + " with tag : " + imageTag);
+        }
+
+        final boolean[] dockerStatusCheck = new boolean[1];
+        dockerStatusCheck[0] = true;
```
Comment (amalkasubasinghe) [inline modules/components/org.wso2.appcloud.core/src/main/java/org/wso2/appcloud/core/docker/DockerClient.java]: shall we make it false and make it true only when it is success??

File: modules/jaggeryapps/appmgt/src/site/themes/default/templates/application/template.jag (hunk 2 above)
Comment (mcvidanagama) [inline modules/jaggeryapps/appmgt/src/site/themes/default/templates/application/template.jag]: This is no longer in the code

File: modules/jaggeryapps/appmgt/src/site/themes/default/templates/home/js/home.js (hunk 1 above)
Comment (mcvidanagama) [inline modules/jaggeryapps/appmgt/src/site/themes/default/templates/home/js/home.js]: No, custom is a variable name not a string

File: modules/jaggeryapps/appmgt/src/modules/application/application.jag (hunk 4 above)
Comment (mcvidanagama) [inline modules/jaggeryapps/appmgt/src/modules/application/application.jag]: Will work on this

File: modules/dbscripts/appcloud.sql (hunk 5 above)
Comment (mcvidanagama) [inline modules/dbscripts/appcloud.sql]: Why not, Users can use any of the java version with their image. Therefore I think java version is also custom as the OS. Isnt it ?

File: modules/components/org.wso2.appcloud.core/src/main/java/org/wso2/appcloud/core/docker/DockerClient.java (hunk 7 above)
Comment (mcvidanagama) [inline modules/components/org.wso2.appcloud.core/src/main/java/org/wso2/appcloud/core/docker/DockerClient.java]: Change done

File: modules/components/org.wso2.appcloud.core/src/main/java/org/wso2/appcloud/core/docker/DockerClient.java (hunk 6 above)
Comment (mcvidanagama) [inline modules/components/org.wso2.appcloud.core/src/main/java/org/wso2/appcloud/core/docker/DockerClient.java]: Change done

**Pull request merged**